cpu/esp{32,8266}: model kconfig

leandrolanzieri opened the pull request:
### Contribution description
This models the Kconfig modules for all boards based on ESP32 and ESP8266, I added one board of each CPU to the tests. For now the networking is not included. 6c7ce3d49c6f3b4ad9929147482be3946c2624f2 limits the CI to test only this two boards for now so we can iterate quickly, I'll remove it once everything is green.

### Testing procedure
- Green CI
- Check the modelling and compare `make info-modules` list with and without `TEST_KCONFIG=1`, they should match.

### Issues/PRs references
~~Depends on #17286~~
~~Depends on #17287~~
~~Depends on #17288~~
~~Depends on #17289~~

Part of #16875

gschorcht commented:
@leandrolanzieri The changes of following files seem not to be related to ESPs:
`cpu/native/Kconfig`
`cpu/native/vfs/Kconfig`
`drivers/ili9341/Kconfig`
`makefiles/features_modules.inc.mk`
`sys/Kconfig`
`sys/vfs/Kconfig`
Are they included by mistake?

For the following file I'm not sure whether it has to do with the `esp32-heltec-lora32-v2` board.
`drivers/sx126x/Kconfig`

leandrolanzieri commented:
> @leandrolanzieri The changes of following files seem not to be related to ESPs: `cpu/native/Kconfig` `cpu/native/vfs/Kconfig` `drivers/ili9341/Kconfig` `makefiles/features_modules.inc.mk` `sys/Kconfig` `sys/vfs/Kconfig` Are they included by mistake?
> 
> For the following file I'm not sure whether it has to do with the `esp32-heltec-lora32-v2` board. `drivers/sx126x/Kconfig`

They were not included by mistake but needed changes to model these boards. I can try to split them if you think it makes it easier.

gschorcht commented:
> > @leandrolanzieri The changes of following files seem not to be related to ESPs: `cpu/native/Kconfig` `cpu/native/vfs/Kconfig` `drivers/ili9341/Kconfig` `makefiles/features_modules.inc.mk` `sys/Kconfig` `sys/vfs/Kconfig` Are they included by mistake?
> 
> They were not included by mistake but needed changes to model these boards. I can try to split them if you think it makes it easier.

If they are needed not only by these boards, but also by others and were simply forgotten, we should provide them with a separate PR first.

leandrolanzieri commented:
> > > @leandrolanzieri The changes of following files seem not to be related to ESPs: `cpu/native/Kconfig` `cpu/native/vfs/Kconfig` `drivers/ili9341/Kconfig` `makefiles/features_modules.inc.mk` `sys/Kconfig` `sys/vfs/Kconfig` Are they included by mistake?
> > 
> > 
> > They were not included by mistake but needed changes to model these boards. I can try to split them if you think it makes it easier.
> 
> If they are needed not only by these boards, but also by others and were simply forgotten, we should provide them with a separate PR first.

@gschorcht the other files were not 'simply forgotten'. We are currently in the middle of this migration, so many times we move forward by modelling whole boards/cpus which need peripheral work, that's why those changes were included here. It makes it easier to test and stresses less our CI. Nevertheless, I split this PR in #17286,  #17287, #17288, #17289, which are dependencies for this one now.

leandrolanzieri commented on boards/esp32-wrover-kit/esp32-wrover-kit.config:
```diff
@@ -0,0 +1,2 @@
+# Sets up configuration for openocd
+CONFIG_MODULE_ESP_JTAG=y
```
But it is enabled by default
https://github.com/RIOT-OS/RIOT/blob/c84a40abc4511f2bc5b55f0db8e708432e32d161/boards/esp32-wrover-kit/Makefile.dep#L5-L6

gschorcht replied:
Indeed, you are right. We have discussed it in https://github.com/RIOT-OS/RIOT/pull/16911#issuecomment-948841846. It will still flash with `esptool` unless `PROGRAMMER` is set to `openocd`. I have forgotten this already. We can leave it as is.

gschorcht replied:
Another question. Why did you model this only setting within a separate file? Wouldn't it be enough to use
```
select MODULE_ESP_JTAG
```
here https://github.com/RIOT-OS/RIOT/blob/6c7ce3d49c6f3b4ad9929147482be3946c2624f2/boards/esp32-wrover-kit/Kconfig#L10-L14

leandrolanzieri replied:
No, `select` forces a module to be `y`, this is configurable by the user (AFAIU), so by using a `.config` file we provide a default configuration.

gschorcht replied:
OK, that is `.config` corresponds always to `Makefile.default`.

leandrolanzieri replied:
Somehow yes. But one can only configure visible symbols this way (i.e., symbols that the user would be able to configure via menuconfig).

gschorcht commented on cpu/esp32/Makefile.dep:
```diff
@@ -79,11 +75,6 @@ ifneq (,$(filter mtd,$(USEMODULE)))
   USEMODULE += esp_idf_spi_flash
 endif
 
-ifneq (,$(filter ndn-riot,$(USEPKG)))
```
You removed this. Isn't this required anymore?

leandrolanzieri replied:
I found this dependency strange that this dependency was only needed for this CPU. I think I tried finding out in the initial introduction why this was required but I could not find it. Why is this needed for esp32?

gschorcht replied:
> I found it strange that this dependency was only needed for this CPU.

Yes, but ESPs are the only CPUs that have an WiFi interface and use `wpa_supplicant` which comes with its own `crypto` module :wink:

When we added ESP32 support initially, there were symbol conflicts of `sys/crypto`/`sys/hashes` with `wpa_supplicant/crypto`. So we had to avoid that both `crypto` modules were used at the same time.

This fragment was only a part of the dependency resolution. `cpu/esp32/Makefile.dep` also included the following: https://github.com/RIOT-OS/RIOT/blob/3ac99877ac2939b087e7b639041c08a4ad9071fb/cpu/esp32/Makefile.dep#L29-L38.

With PR #11994 we solved the conflict by renaming the symbols in `wpa_supplicant` to something `wpa_<symbol>`. The part of the dependencies above were just forgotten to remove. So we can do this with this PR :+1:

leandrolanzieri replied:
Ok, thanks for clarifying !

gschorcht commented on boards/common/esp32/Kconfig:
```diff
@@ -10,3 +10,13 @@ config BOARD_COMMON_ESP32
+
+    select HAVE_SAUL_GPIO
+
+config MODULE_BOARDS_COMMON_ESP32
```
Why is this one selectable while the that one in `boards/common/esp8266/Kconfig` is not? The structure of `Makefile.dep` is the same for both.

leandrolanzieri replied:
What do you mean by selectable? It has no prompt as well, the user can't select them, both are hidden symbols

leandrolanzieri replied:
No particular reason, I probably did that on two separate days. Being a hidden symbol that is always enabled both approaches are pretty much the same. I changed to the `default y` to use less `select`. 190e76b439b98e89562b8454ec161a02beec55c0

gschorcht commented on cpu/esp32/Kconfig:
```diff
@@ -93,4 +100,22 @@ config CPU_MODEL
+
+config MODULE_ESP_SPI_RAM
+    bool "SPI RAM support"
+    select MODULE_ESP_IDF_HEAP
```
```suggestion
    depends on HAS_ESP_SPI_RAM
    select MODULE_ESP_IDF_HEAP
```
This option should only be available for boards that provide SPI RAM. I know we didn't make this difference in `Makefile.dep`s by using `FEATURES_PROVIDED` and `FEATURES_REQUIRED`, but with the introduction of Kconfig we could solve this problem. The question is whether we should also enhance the handling of this feature  in `Makefile.dep`s with this change.

gschorcht replied:
BTW, feature `HAS_ESP_SPI_RAM` could be modelled using `CPU_MODEL_ESP32_WROVER*` because SPI RAM is a feature of WROVER modules and all board that use the WROVER module also have SPI RAM. The only exception is the `esp32-ttgo-t-beam` board which has a ESP SoC and an external SPI RAM connected to it.

leandrolanzieri replied:
Sure, makes sense. I'll try to add it. As we are currently comparing to Makefile features should match, we'll have to add it on both places.

leandrolanzieri replied:
The feature is refactored in Makefile here: b2b370669d123187a8d61dbc721c38f6ca933a3c. The dependency added here 446c7824ee5dd401478ac458cb0e4bfba46322ec. And for Kconfig here 7194df0726a9296d4dbd6555b5b8323cd3a473f5

gschorcht commented on cpu/esp_common/include/log/log_module.h:
```diff
@@ -26,15 +26,6 @@ extern "C" {
 #include <stdio.h>
 #include "esp_common_log.h"
 
-#ifdef MODULE_LOG_PRINTFNOFORMAT
```
What is the reason for changing the approach? I don't see any advantage to handle it as pseudomodule. ESPs require their own log implementation for compatibility and consistency reasons with the SDK and SDK libraries. It's a bit tricky but it works. Removing this part breaks the compatibility with module `log_printnoformat` so that `tests/log_printfnoformat` isn't compilable anymore.

leandrolanzieri replied:
I don't think the current solution is good. The implementation of `log_printfnoformat` is being duplicated here for no reason, only to always include this header file. The way log backends are provided is explicitly defined on its documentation:
https://github.com/RIOT-OS/RIOT/blob/44ebc38caca3ce24914a6b543fb34703a304e660/core/include/log.h#L20-L27

> Removing this part breaks the compatibility with module log_printnoformat so that tests/log_printfnoformat isn't compilable anymore.

What do you mean by breaks compatibility?

In my opinion, if ESP needs a special logging which, for some reason, is not compatible with the existing ones, then one should not be able to use them. Another way is to provide their ESP version (e.g. `log_esp_prinfnoformat` or so), but always using modules, otherwise all this is really hard to see and maintain.

gschorcht replied:
> What do you mean by breaks compatibility?

With your changes, `tests/log_printfnoformat` is no longer compilable. This means that we either blacklist ESP boards for this test or define a dependency that we do not use the `esp_common/log` module if the `log_printfnoformat` module is used.

My guess is that the `log_printfnoformat` module exists to allow the user to disable `printf` logging to reduce ROM/RAM size. I think we should keep this option for ESP boards as well, even if the SDK library outputs are then no longer consistent and contain `printf` formatting sequences.

leandrolanzieri replied:
> we do not use the esp_common/log module if the log_printfnoformat module is used.

I'm thinking we can do this using `Makefile.default`, right? ESP always tries to enable that logging backend by default and when specifically configured it is disabled.

gschorcht replied:
If this work, I'm fine with it.

leandrolanzieri replied:
- A sanity check for dependencies is added in b9bea0b8b2749c18fb219d4bb50f909642e58bfc.
- The default is set in fd2255f45d3df7e6e25a90fb66fe18e8e26f4107

gschorcht replied:
Works as expected :+1:

gschorcht commented on cpu/esp32/Kconfig:
```diff
@@ -93,4 +100,22 @@ config CPU_MODEL
 config CPU
     default "esp32" if CPU_FAM_ESP32
 
+menu "ESP32 configurations"
```
Not part of this PR. Do you mind to move
```
config HAS_ESP_WIFI_ENTERPRISE
    bool
    help
        Indicates that the platform supports WPA2 enterprise mode for the WiFi
        interface.
```
to `cpu/esp_common/Kconfig`? It's not a special feature of ESP32 and with a newer ESP8266 it will also be a feature of ESP8266. So the definition of the feature should be in `esp_common` and the CPU should only select the feature.

leandrolanzieri replied:
Ok, I moved it in a2700de6ad7dc391317c6d7dd159ae99781cde81

gschorcht commented on cpu/esp32/periph/Kconfig.i2c:
```diff
@@ -0,0 +1,36 @@
+    bool "I2C implementation"
+    depends on MODULE_PERIPH_I2C
+    # necessary to fix driver initialization on esp32
+    default MODULE_ESP_I2C_HW if MODULE_AIP31068
```
This dependency isn't clear, where does it come frome? AIP31068 is a liquid crystal I2C/SPI LCD display.

gschorcht replied:
The default should always be the software implementation.

leandrolanzieri replied:
It comes from here:
https://github.com/RIOT-OS/RIOT/blob/44ebc38caca3ce24914a6b543fb34703a304e660/drivers/aip31068/Makefile.dep#L5-L8

I also don't like this, but we are trying to match the Makefile version. @HendrikVE  why is this dependency needed?

gschorcht replied:
Hm, @HendrikVE does this mean that AIP31068 is only working with the hardware implementation of I2C?

gschorcht replied:
BTW, out of topic, I have open PR #12665 since two years which simply extends the existing HD44170 driver by the I2C IO expander PCF857x. I didn't take a look into PR #14966 in detail.

HendrikVE replied:
I'm not entirely sure what the problem was back then. I don't have the hardware with me today so I'll try it again at home this evening and report back

HendrikVE replied:
@gschorcht I did a quick test and I can confirm that this problem still exists. When using `esp_i2c_hw` then the initialization always work. But without it it will fail in 80% of the cases with `rc = -6`. 

It happens here:
https://github.com/RIOT-OS/RIOT/blob/c84a40abc4511f2bc5b55f0db8e708432e32d161/drivers/aip31068/aip31068.c#L145
It is caused by:
https://github.com/RIOT-OS/RIOT/blob/207964265a459579ef38781be8945366d3cc88a4/cpu/esp_common/periph/i2c_sw.c#L365

gschorcht replied:
This is really strange, I use the I2C software implementation on ESP32 and EP8266 very extensively with a large number of different devices and never had any problems. My impression was rather that the hardware implementation is a bit unstable, the I2C hardware sometimes gets stuck, so there is an extra bus `_i2c_reset_hw` cycle in each `i2c_acquire` and `i2c_release` which generates 20 clock cycles while the SDA line is held LOW, which creates a noticeable delay in each transfer.

Without having the hardware I can't initialize the problem with logic analyzer but it would be very interesting why the I2C software implementation returns with that error in 80 % of the cases.

gschorcht replied:
Ok, I have tested the LCD module with the I2C software implementation. On ESP32 I can reproduce that very often it is not possible to communicate with the AIP31068. But the same software implementation is used for ESP8266 where it works reliably. The only difference between ESP32 and ESP8266 is that the GPIOs are used in a different mode for timing reasons. Maybe that makes the difference.

Anyway, using the same mechanism to clear the bus during I2C initialization (sending 10 clock pulses while SDA is low) solves this problem for the software implementation on ESP32. Obviously, the AIP31068 is a special sensitive. I use ESP32 very intensively with a large number of different I2C devices and have never had such problems.

Finally, I would prefer to remove this special case. I will provide a fix for the software implementation for ESP32.

BTW, my module is from Waveshare which has an additional RGB backlight control using a PCA9633DP2 which has to controlled too, otherwise you don't have backlight as in my case. Maybe we should extend your example by the option to enable module `pca9633` to control the backlight.

leandrolanzieri replied:
With #17346 in, I pushed e6126904389b0a58c24c0a14fdd1e74eb479064b which removes this strange dependency. Thanks @gschorcht !

gschorcht commented on cpu/esp32/vendor/esp-idf/driver/Kconfig:
```diff
@@ -0,0 +1,11 @@
+config MODULE_ESP_IDF_DRIVER
+    bool
+    depends on TEST_KCONFIG
+    depends on HAS_ARCH_ESP
```
Here an for all others in subdirectories of `cpu/esp32/vendor/esp-idf`
```suggestion
    depends on HAS_ARCH_ESP32
```
The vendor code is specific for ESP32.

leandrolanzieri replied:
Now it depends on `MODULE_ESP_IDF`, which is already depending on the arch feature

gschorcht commented on cpu/esp8266/vendor/esp-idf/esp8266/source/Kconfig:
```diff
@@ -0,0 +1,11 @@
+config MODULE_ESP_IDF_ESP8266
+    bool
+    depends on TEST_KCONFIG
+    depends on MODULE_ESP_IDF
```
```suggestion
    depends on HAS_ARCH_ESP8266
    depends on MODULE_ESP_IDF
```

leandrolanzieri replied:
Added in e38f6154042eae0f189239e4e10a847321edae65

gschorcht commented on cpu/esp8266/vendor/esp-idf/Kconfig:
```diff
@@ -0,0 +1,18 @@
+
+rsource "esp8266/source/Kconfig"
+rsource "heap/src/Kconfig"
+rsource "nvs_flash/Kconfig"
```
Module is not compiled in `nvs_flash` but in `nvs_flash/src`. I guess the `Kconfig` has to be in that directory to use the right makefile. 
```suggestion
rsource "nvs_flash/src/Kconfig"
```

leandrolanzieri replied:
Hmm now that you point this out I'm confused. Why is this makefile there, with the module name?

https://github.com/RIOT-OS/RIOT/blob/master/cpu/esp8266/vendor/esp-idf/nvs_flash/Makefile

gschorcht replied:
It's obviously a leftover. Compilation works without this Makefile. You could remove it with this PR.

leandrolanzieri replied:
Will do.

leandrolanzieri replied:
Makefile removed here 4b09f491ca921f35503d493d99085f3814c48ff0.

gschorcht commented on cpu/esp8266/vendor/esp-idf/nvs_flash/Kconfig:
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2021 HAW Hamburg
```
Should be moved to `nvs_flash/src`.

leandrolanzieri replied:
Moved in e38f6154042eae0f189239e4e10a847321edae65

gschorcht commented on cpu/esp32/vendor/esp-idf/Kconfig:
```diff
@@ -0,0 +1,25 @@
+rsource "heap/Kconfig"
+rsource "nvs_flash/Kconfig"
+rsource "soc/Kconfig"
+rsource "spi_flash/Kconfig"
```
Is `wpa_supplicant/Kconfig` missing by intention? I should be almost if not the same as for ESP8266.

leandrolanzieri replied:
I didn't yet get to the networking part, as that is coming later for most of cpus and drivers, due to lack of modelling of related modules, (e.g. netdev).

gschorcht replied:
OK, I was just wondering because you did it for EPS8266.

gschorcht replied:
Do you mind to add at least the module definition for `wpa_supplicant` as for ESP8266 so that the vendor libraries are consistent and we can later rely on that when we model the network interfaces like `esp_now` and `esp_wifi`?

leandrolanzieri replied:
I added all `wpa_supplicant` modules in c364ca7185b0663ba513ff2ac838e3d7c86502f6

leandrolanzieri commented on .murdock:
```diff
@@ -207,7 +209,7 @@ get_supported_boards() {
     do
         echo $board
-    done | $(_greplist $BOARDS)
+    done | $(_greplist $BOARDS) | $(_greplist esp32-heltec-lora32-v2 esp8266-esp-12x)
```
This is just temporary, to limit the boards that are being build for rapid iteration on the CI while making changes on the PR, it is a `REMOVEME:` commit (it blocks merging), I'll remove it once we are done.

gschorcht replied:
OK, I'm wondering whether it would make sense also to add `esp32-wrover-kit` because it is the one of the most featured board and I already figured out a missing dependency, see https://github.com/RIOT-OS/RIOT/pull/17232#discussion_r759054428.

leandrolanzieri replied:
Sure, I'll add all of them

leandrolanzieri replied:
All boards added in 78aff83565141d764a561f55505f5f32b0dda927

gschorcht commented on boards/esp32-wrover-kit/Kconfig:
```diff
@@ -20,6 +20,7 @@ config BOARD_ESP32_WROVER_KIT
     select HAS_PERIPH_SPI
     select HAS_SDCARD_SPI
-    select MODULE_ESP_JTAG
+
```
Please add
```
    select HAS_ESP_JTAG
```
to the list otherwise dependencies fail:
```
[genconfig.py]:ERROR-=> The module MODULE_ESP_JTAG could not be set to y.
[genconfig.py]:ERROR-   Check the following unmet dependencies: HAS_ESP_JTAG (=n)
```

leandrolanzieri replied:
The feature is missing on both makefile and Kconfig, but it is not failing on the Makefile-based dependency resolution because the feature is being required on the wrong file
https://github.com/RIOT-OS/RIOT/blob/44ebc38caca3ce24914a6b543fb34703a304e660/cpu/esp32/Makefile.include#L53-L55

I'll move it to the `Makefile.dep` and provide the feature on both Kconfig and Makefile.

leandrolanzieri replied:
JTAG dependency fixed in Makefile here bff8cc88b174e4201ff56ce9ca006842ab35fadc. Now the `esp32-wrover-kit` provides the feature (013490280b7c9233630529fe3270f03e6a4a16c1). And for the Kconfig counterpart: 9d0e0d94016e0bed4ab14f492a8df2f05d9d6d5e

gschorcht commented on cpu/esp_common/esp_common.config:
```diff
@@ -0,0 +1,8 @@
+CONFIG_MODULE_RANDOM=y
+CONFIG_MODULE_PRNG_TINYMT32=y
```
Where does this come from, I couldn't find it in any `Makefile*` for ESPs boards and the CPUs.

leandrolanzieri replied:
ESP brings random by default. That module provides multiple backends. While in Makefile checking for HWRNG and default to it when possible is not easy to model, Kconfig provides this as a choice (see [here](https://github.com/RIOT-OS/RIOT/blob/44ebc38caca3ce24914a6b543fb34703a304e660/sys/random/Kconfig)). As Kconfig would default to HWRNG but Makefile does not, we are defaulting to a software implementation whenever it is used. See a discussion on this in https://github.com/RIOT-OS/RIOT/pull/17188

gschorcht replied:
Ok, it is only for compaibility of kconfig with Mafiles. Will this be changed once PR #16732 is merged, if it is merged? From my point of view it makes sense to use the HWRNG whenever it is possible.

leandrolanzieri replied:
With #17188 merged now the default is `MODULE_PRNG_MUSL_LCG`, I'll adapt

leandrolanzieri commented on boards/common/esp32/Kconfig:
```diff
@@ -10,3 +10,13 @@ config BOARD_COMMON_ESP32
+    bool
+    depends on TEST_KCONFIG
+    depends on BOARD_COMMON_ESP32
+    default y
```
Added be142fd29c8a9d4bc1d8de82a46d664b624d579c

gschorcht commented on boards/common/esp8266/Kconfig:
```diff
@@ -15,3 +15,13 @@ config BOARD_COMMON_ESP8266
+config MODULE_BOARDS_COMMON_ESP8266
+    bool
+    depends on TEST_KCONFIG
+    depends on HAS_ARCH_ESP8266
```
```suggestion
    depends on BOARD_COMMON_ESP8266
    depends on HAS_ARCH_ESP8266
```
Doesn't it require `BOARD_COMMON_ESP8266`. I'm just wondering because this is different to the definition for ESP32.

leandrolanzieri replied:
I can add it, sure. Keep in mind that the only way for this file to be loaded is that we are on an esp8266 board, otherwise the symbol is not there. For consistency it makes sense to add it.

leandrolanzieri replied:
Changed the dependency in 9e66f139cc25e71ff308710012fbcccd6893e842

gschorcht commented on boards/common/esp8266/Kconfig:
```diff
@@ -21,7 +21,7 @@ config BOARD_COMMON_ESP8266
     bool
     depends on TEST_KCONFIG
-    depends on HAS_ARCH_ESP8266
+    depends on BOARD_COMMON_ESP8266
```
```
    depends on HAS_ARCH_ESP8266
```
was correct because `BOARD_COMMON_ESP8266` doesn't depend on `HAS_ARCH_ESP8266` if I'm right. My suggestion was to add
```
depends on BOARD_COMMON_ESP8266
```
as for ESP32 to have the dependencies defined by `BOARD_COMMON_ESP8266`. Or am I wrong?

leandrolanzieri replied:
Hmm yeah. I re-added the dependency on the feature fc0c3d83c27d39b674f03aa0064b6d1d92c9c582

gschorcht commented:
@leandrolanzieri Out of topic. When I was crawling through configurations with `TEST_KCONFIG=1`, it was really difficult to find a certain modules because the order depend on module name while the short description is shown, e.g. in `System`,
```
[ ] A simple CLI progress bar
[ ] UNIX like ps command
[*] Pseudo-Random Number Generation  --->
[ ] Serial number arithmetic (RFC 1982)
```
Is there something planned which makes it easier to find a certain module? In `Drivers -> Sensor Device Drivers`, it is easy since all driver entries start with the sensor name.

gschorcht commented:
@leandrolanzieri Out of topic. Another thing I have seen is that configurations that depend on the selection of a module in `Drivers - Peripherals drivers` are handled differently. For some peripherals the `Auto initialize` option becomes visible when the according peripheral is selected, for others they are handled as the only entry in a submenu for example:
```
[*] Power Management (PM) peripheral driver
[*]     Auto initialize Power Management (PM) peripheral
```
vs. 
```
[*] RTC peripheral driver  --->
```
Furthermore, for I2C a submenu as well as an option becomes visible.
```
[*] I2C peripheral driver  --->
    I2C implementation (Software)  --->
```
Shouldn't the I2C implementation selection be part of the I2C submenu?

leandrolanzieri commented:
> Is there something planned which makes it easier to find a certain module? In `Drivers -> Sensor Device Drivers`, it is easy since all driver entries start with the sensor name.

There is nothing planned so far. Currently we are trying to model all of them. It would be really nice to start grouping system modules into `menu`s. Regarding the name: we could prefix the prompt with it if it makes it easier.

leandrolanzieri commented:
> @leandrolanzieri Out of topic. Another thing I have seen is that configurations that depend on the selection of a module in `Drivers - Peripherals drivers` are handled differently. For some peripherals the `Auto initialize` option becomes visible when the according peripheral is selected, for others they are handled as the only entry in a submenu for example:
> 
> ```
> [*] Power Management (PM) peripheral driver
> [*]     Auto initialize Power Management (PM) peripheral
> ```
> 
> vs.
> 
> ```
> [*] RTC peripheral driver  --->
> ```
> 


The way this is displayed depends on the position of the symbols. When `menuconfig` is used to define a symbol, and one or more symbols depend on it and are defined right after it they show up in this sort of nested menu. If `menuconfig` is not used to define the symbol (`config` was used instead), but there are one or more symbols defined right after which depend on it, they will be nested but without the submenu view.

> Furthermore, for I2C a submenu as well as an option becomes visible.
> 
> ```
> [*] I2C peripheral driver  --->
>     I2C implementation (Software)  --->
> ```
> 
> Shouldn't the I2C implementation selection be part of the I2C submenu?

This is because the `source` is outside the first `if`. Eventually we can move it inside, I did not test if that would show the CPU-specific options inside the menu.

gschorcht commented:
Unfortunatly, we have also to remodel `periph_rtc` here https://github.com/RIOT-OS/RIOT/blob/61177f1bc942e9afeca537ebf71a34b93fe0db14/cpu/esp32/Kconfig#L19 and here https://github.com/RIOT-OS/RIOT/blob/61177f1bc942e9afeca537ebf71a34b93fe0db14/cpu/esp8266/Kconfig#L17 Although `periph_rtc` is declared as a feature in `Makefile.features`, the ESP implementation no longer supports `periph_rtc`. The reason is that the timers are not RTCs but RTTs. Therefore ESP implementations provide `periph_rtt` and use the RTC emulation on top of RTT (`rtt_rtc`) if `periph_rtc` is used https://github.com/RIOT-OS/RIOT/blob/61177f1bc942e9afeca537ebf71a34b93fe0db14/cpu/esp32/Makefile.dep#L82-L84 The RTC feature declarations are just leftover that didn't lead to any problem in Makefile dependencies.

I have tried to add
```diff 
diff --git a/cpu/esp32/Kconfig b/cpu/esp32/Kconfig
index 4590cc3ebd..c19a0cdefb 100644
--- a/cpu/esp32/Kconfig
+++ b/cpu/esp32/Kconfig
@@ -16,12 +16,12 @@ config CPU_FAM_ESP32
     select HAS_CPU_ESP32
     select HAS_ESP_WIFI_ENTERPRISE
     select HAS_PERIPH_ADC_CTRL
-    select HAS_PERIPH_RTC
     select HAS_PERIPH_RTT
     select HAS_PERIPH_RTT_SET_COUNTER
     select HAS_PERIPH_RTT_OVERFLOW
 
     select MODULE_PERIPH_RTT if HAS_PERIPH_RTT && MODULE_PM_LAYERED
+    select MODULE_RTT_RTC if HAS_PERIPH_RTT && MODULE_PERIPH_RTC
     select MODULE_PS if MODULE_SHELL
     select MODULE_ESP_IDF_DRIVER if TEST_KCONFIG
     select MODULE_ESP_IDF_ESP32 if TEST_KCONFIG
```
but I get dependency loop errors when I try to use
```
TEST_KCONFIG=1 make BOARD=esp32-wroom-32 -C tests/periph_rtc menuconfig
```

leandrolanzieri commented:
> but I get dependency loop errors when I try to use

Yes, this come from the `select HAS_PERIPH_RTC` of  `MODULE_RTT_RTC`. I guess we should get rid of it now, as the common use for it is not actually as I envisioned. Instead, platforms without RTC announce having one and select this. I'll remove the feature selection. 

A bit unrelated to the particular PR, but for some background on modelling modules in Kconfig: There is some initial discussion in  #15429. In a nutshell: it is recommended to use `depends on` instead of `select` for visible symbols, but makes configurations more verbose and not so 'automatic'. In a more conservative way of modelling, where modules are enabled bottom-up, a module that uses the RTC API would not care who provides this, only that it is there. An example could be:

```
config MODULE_MY_DRIVER
    bool "My Driver support"
    depends on RTC

# [...]
# Include these modules as choice options as they are mutually exclusive

choice RTC_PROVIDERS
    bool "RTC support"
    optional

config MODULE_PERIPH_RTC
    bool "RTC peripheral driver"
    depends on HAS_PERIPH_RTC
    select RTC

config MODULE_RTT_RTC
    bool "RTT-based RTC implementation"
    depends on HAS_PERIPH_RTT
    select RTC

endchoice

config RTC
    bool
    help
      Indicates that RTC support is provided.
```

But of course the configuration file could not simply enable the driver, as it needs to know which RTC provider can be enabled on the platform. By adding an intermediate symbol, the situation can be improved though:

```
config MODULE_MY_DRIVER
    bool "My Driver support"
    depends on RTC

# [...]

menuconfig RTC
    bool "RTC support"
    depends on HAS_PERIPH_RTC || HAS_PERIPH_RTT

choice RTC_PROVIDERS
    bool "Choose a backend"
    depends on RTC

config MODULE_PERIPH_RTC
    bool "RTC peripheral driver"
    depends on HAS_PERIPH_RTC

config MODULE_RTT_RTC
    bool "RTT-based RTC implementation"
    depends on HAS_PERIPH_RTT

endchoice
```
Now, a configuration file to enable the driver can look like the following, independent of the platform. New backends can be added and the driver symbol does not change its dependencies.

```makefile
CONFIG_MODULE_MY_DRIVER=y
CONFIG_RTC=y
```

leandrolanzieri commented:
> Unfortunatly, we have also to remodel `periph_rtc` here

Done in ddd9453d95, c1bb3c95e514d79c0269653262c4dea7113beb95 and 7ad4db28916946dbd7168996c9073c2138739583

gschorcht commented:
> A bit unrelated to the particular PR, but for some background on modelling modules in Kconfig: There is some initial discussion in #15429. In a nutshell: it is recommended to use `depends on` instead of `select` for visible symbols, but makes configurations more verbose and not so 'automatic'.

I am not sure, but I think I understood. But in the case of the `benchmark` module it was correct to use `select` even if `MODULE_ZTIMER` is visible because otherwise it could not be used, right?

Since the use of the `ztimer_usec` module is a prerequisite to be able to use `benchmark` at all, I had even considered if
```
select ZTIMER_USEC`
```
would be correct instead of
```
select MODULE_ZTIMER
select MODULE_ZTIMER_USEC
```
which activates both modules and the usec clock so that the user can't change them.

leandrolanzieri commented:
> But in the case of the `benchmark` module it was correct to use `select` even if `MODULE_ZTIMER` is visible because otherwise it could not be used, right?

Yes, given the wide use of ztimer we are trying the `select` approach with it, so it was a correct change.

gschorcht commented:
> > But in the case of the `benchmark` module it was correct to use `select` even if `MODULE_ZTIMER` is visible because otherwise it could not be used, right?
> 
> Yes, given the wide use of ztimer we are trying the `select` approach with it, so it was a correct change.

The last question, would it have been better to use `select ZTIMER_USEC` instead of `select MODULE_ZTIMER` + `select MODULE_ZTIMER_USEC`?

leandrolanzieri commented:
> The last question, would it have been better to use `select ZTIMER_USEC` instead of `select MODULE_ZTIMER` + `select MODULE_ZTIMER_USEC`?

When you did it no, the approach was fine. After https://github.com/RIOT-OS/RIOT/pull/17307 yes, that's why I introduced the change for benchmark there.

gschorcht commented:
`Kconfig` files in `cpu/esp*/vendor/esp-idf/*` are still not consistent. We have for example:

[cpu/esp32/vendor/esp-idf/driver/Kconfig](https://github.com/RIOT-OS/RIOT/pull/17232/files#diff-e4fdb0b78787f8556ebcef36652d1744ad2b9b36b1aef342c5573d69288733f4)
```
    depends on TEST_KCONFIG
    depends on MODULE_ESP_IDF
```
[cpu/esp32/vendor/esp-idf/esp32/Kconfig](https://github.com/RIOT-OS/RIOT/pull/17232/files#diff-d1741c9571125f833eb21a9b836463a4abbecc7289600e0b2bc5533ad7ef894c)
```
    depends on TEST_KCONFIG
    depends on HAS_ARCH_ESP
```
[cpu/esp32/vendor/esp-idf/heap/Kconfig](https://github.com/RIOT-OS/RIOT/pull/17232/files#diff-57ef55a8150ab50608623453fdfce853c9f07e1f2a4ee9a693c14bac15af7f7e)
```
    depends on TEST_KCONFIG
    depends on HAS_ARCH_ESP32
```
[cpu/esp32/vendor/esp-idf/wpa_supplicant/port/Kconfig](https://github.com/RIOT-OS/RIOT/pull/17232/files#diff-ce3f4392e37ab017c5d9e10b7745165e1e18fcd7cd1ab6b6d83838ade5a1f344)
```
   depends on TEST_KCONFIG
   depends on HAS_ARCH_ESP32
   depends on MODULE_ESP_IDF
```
[cpu/esp8266/vendor/esp-idf/esp8266/source/Kconfig](https://github.com/RIOT-OS/RIOT/pull/17232/files#diff-2032869c8b0212400722aa8efab4f7a8e2909d8ba7198e0895ca471399c84f40)
```
    depends on TEST_KCONFIG
    depends on MODULE_ESP_IDF
    depends on HAS_ARCH_ESP8266
```
I hope I am not too nitpicky, but we should use a consistent approach. If I am correct, since `MODULE_ESP_IDF` already depends on `HAS_ARCH_ESP32` or `HAS_ARCH_ESP8266`, it should be sufficient to use `depend on MODULE_ESP_IDF` in all cases.

leandrolanzieri commented:
> `Kconfig` files in `cpu/esp*/vendor/esp-idf/*` are still not consistent. We have for example:

Thanks for the thorough review. I think I addressed this for both 32 and 8266 in  52d3c71c24114884f77a517fef450844d7e98b95 and 52d3c71c24114884f77a517fef450844d7e98b95

gschorcht commented on cpu/esp32/vendor/esp-idf/wpa_supplicant/port/Kconfig:
```diff
@@ -0,0 +1,14 @@
+config MODULE_ESP_IDF_WPA_SUPPLICANT_PORT
+    bool
+    depends on TEST_KCONFIG
+    depends on HAS_ARCH_ESP32
```
```suggestion
```
This file was forgotten.

leandrolanzieri replied:
Fixed 0caa902cddc17a52377716db5376a4a561e72eb7

gschorcht commented:
Please squash. Once the other PRs are merged we can give a try in Murdock.

leandrolanzieri commented:
All dependencies have been merged. I rebased to master now.

leandrolanzieri commented:
I have rebased to resolve the conflicts on `.murdock`.

gschorcht commented on makefiles/features_modules.inc.mk:
```diff
@@ -16,8 +16,13 @@ ifneq (,$(filter periph_init, $(USEMODULE)))
     periph_common \
     periph_flexcomm \
     periph_gpio_mux \
+    periph_i2c_hw \
```
Is this something we need in general or just specifically for ESPs?

leandrolanzieri replied:
Well, the list is global, but these modules appear only for ESPs. It is there just to avoid unnecessary `periph_init_` modules automatically derived.

gschorcht commented on cpu/esp32/Kconfig:
```diff
@@ -29,14 +37,17 @@ config CPU_MODEL_ESP32_WROOM_32
 config CPU_MODEL_ESP32_WROVER
```
`HAS_ESP_SPI_RAM ` is now a feature that is already provided by `ESP32_WROVER*` modules. Thus
```
    select HAS_ESP_SPI_RAM
```
can be removed from board definitions that use these modules as CPU:
```
boards/esp32-ethernet-kit-v1_0/Kconfig
boards/esp32-ethernet-kit-v1_1/Kconfig
boards/esp32-ethernet-kit-v1_2/Kconfig
boards/esp32-wemos-lolin-d32-pro/Kconfig
boards/esp32-wrover-kit/Kconfig
```
We need it in `boards/esp32-ttgo-t-beam/Kconfig` because it is not provided by the CPU module but as separate module. But, the second `select HAS_ESP_SPI_RAM` should be removed.

gschorcht commented on cpu/esp32/Kconfig:
```diff
@@ -29,14 +37,17 @@ config CPU_MODEL_ESP32_WROOM_32
 config CPU_MODEL_ESP32_WROVER
     bool
     select CPU_FAM_ESP32
+    select HAS_ESP_SPI_RAM
```
HAS_ESP_SPI_RAM is now a feature that is already provided by ESP32_WROVER* modules. Thus
```
select HAS_ESP_SPI_RAM
```
can be removed from board definitions that use these modules as CPU:
```
boards/esp32-ethernet-kit-v1_0/Kconfig
boards/esp32-ethernet-kit-v1_1/Kconfig
boards/esp32-ethernet-kit-v1_2/Kconfig
boards/esp32-wemos-lolin-d32-pro/Kconfig
boards/esp32-wrover-kit/Kconfig
```
We need it in `boards/esp32-ttgo-t-beam/Kconfig` because it is not provided by the CPU module but as separate module. But, the second `select HAS_ESP_SPI_RAM` should be removed.

leandrolanzieri replied:
Done in a6e3ace6a98f0ba587e4a4ebd48e108408af6ec9 and 84370c4cc45f393e81d97db8701962e1be804a00

gschorcht commented:
Regarding the failed compilation jobs, there is indeed a difference between compiling with Kconfig and make dependencies, although the list of modules used is the same.

gschorcht commented:
The difference seems to be in using wrong functions for heap in link command if `TEST_KCONFIG=1` is used. With makefile dependencies, following wrapper symbols are used for heap functions in link command if the `esp_idf_heap` module is used:
```
-Wl,-wrap,_malloc_r -Wl,-wrap,_calloc_r -Wl,-wrap,_realloc_r -Wl,-wrap,_free_r
```
They are not used with Kconfig.

These dependencies are defined here
https://github.com/RIOT-OS/RIOT/blob/1efc53b59895162dd0b95db2d7b74a20dbeabd4e/cpu/esp_common/Makefile.dep#L86-L91 and https://github.com/RIOT-OS/RIOT/blob/1efc53b59895162dd0b95db2d7b74a20dbeabd4e/cpu/esp_common/Makefile.include#L91-L95 I don't rember why `-Wl,-wrap,_calloc_r` is defined at two different places. I guess `Makefile.include` is the right place for all these special options.

I have no idea why it doesn't work correctly with `TEST_KCONFIG=1`. `info-modules` and `info-debug-variable-USEMODULE` targets returns the same used modules as for makefile dependencies.

gschorcht commented:
Another difference I saw is that with makefile dependencies `libstdc++.a` is linked by default, with `TEST_KCONFIG` not.

leandrolanzieri commented:
> I don't rember why `-Wl,-wrap,_calloc_r` is defined at two different places. I guess `Makefile.include` is the right place for all these special options.
> I have no idea why it doesn't work correctly with `TEST_KCONFIG=1`. `info-modules` and `info-debug-variable-USEMODULE` targets returns the same used modules as for makefile dependencies.

This is probably not working in Kconfig because `Makefile.dep` are not parsed, I'll try moving them to `Makefile.include`.

leandrolanzieri commented:
> Another difference I saw is that with makefile dependencies `libstdc++.a` is linked by default, with `TEST_KCONFIG` not.

I think I found the issue: the library is added based on a check of `FEATURES_USED`, I'll try to change this to a module instead.

leandrolanzieri commented:
>> I don't rember why `-Wl,-wrap,_calloc_r` is defined at two different places. I guess `Makefile.include` is the right place for all >these special options.
>> I have no idea why it doesn't work correctly with `TEST_KCONFIG=1`. `info-modules` and `info-debug-variable->USEMODULE` targets returns the same used modules as for makefile dependencies.
>
>This is probably not working in Kconfig because `Makefile.dep` are not parsed, I'll try moving them to `Makefile.include`.

Moved in 30f027497d8215c9e6ff79e3c9e44442f5cdc32f

leandrolanzieri commented:
> > Another difference I saw is that with makefile dependencies `libstdc++.a` is linked by default, with `TEST_KCONFIG` not.
> 
> I think I found the issue: the library is added based on a check of `FEATURES_USED`, I'll try to change this to a module instead.

Changed in bb4bc865495b0ef9771574e3b0d88f4db0a9ee90. This works because the used feature is turned into a module here:
https://github.com/RIOT-OS/RIOT/blob/228ba83dc6d519efc333bc99aee18909c67b825f/Makefile.dep#L105-L108

leandrolanzieri commented:
I rebased to resolve conflicts.

gschorcht commented:
> I rebased to resolve conflicts.

Lets see whether Murdoc is happy.

gschorcht reviewed (approved):
Looks good now.

leandrolanzieri commented on cpu/esp_common/Makefile.include:
```diff
@@ -91,7 +97,10 @@ LINKFLAGS += -nostdlib -Wl,-gc-sections -Wl,-static
   # use the wrapper functions for calloc to add correct overflow detection missing
   # in the newlib's version.
-  LINKFLAGS += -Wl,-wrap=_calloc_r
+  LINKFLAGS += -Wl,-wrap,_malloc_r
```
@gschorcht are you sure these are needed? Applications are failing to link now that these actually are being applied to the linker command. In current master I only see `-wrap,_malloc_r`. Removing the others prevents linking from failing.

gschorcht replied:
Definitely yes, if module `esp_idf_heap` is used, all newlib heap function have to be mapped to the heap functions of the ESP-IDF.

I will take a look later when I'm back in the office.

gschorcht commented on cpu/esp_common/Makefile.include:
```diff
@@ -91,7 +97,10 @@ LINKFLAGS += -nostdlib -Wl,-gc-sections -Wl,-static
   # use the wrapper functions for calloc to add correct overflow detection missing
   # in the newlib's version.
-  LINKFLAGS += -Wl,-wrap=_calloc_r
+  LINKFLAGS += -Wl,-wrap,_malloc_r
+  LINKFLAGS += -Wl,-wrap,_calloc_r
+  LINKFLAGS += -Wl,-wrap,_realloc_r
+  LINKFLAGS += -Wl,-wrap,_free_r
```
```suggestion
  LINKFLAGS += -Wl,-wrap,_calloc_r
endif

ifneq (,$(filter esp_idf_heap,$(USEMODULE)))
  LINKFLAGS += -Wl,-wrap,_malloc_r
  LINKFLAGS += -Wl,-wrap,_calloc_r
  LINKFLAGS += -Wl,-wrap,_realloc_r
  LINKFLAGS += -Wl,-wrap,_free_r
```
You moved this block from `Makefile.dep` to here. The conditional that was already here was `ifeq` but the conditional in `Makefile.dep` was `ifneq`. That is, `_calloc_r` was wrapped in any case, but the others only if moculde `esp_idf_heap` is used.

So the following should work too:
```diff
--- a/cpu/esp_common/Makefile.include
+++ b/cpu/esp_common/Makefile.include
@@ -94,11 +94,12 @@ LINKFLAGS += -L$(ESP_SDK_DIR)/components/$(CPU)/lib
 endif
 LINKFLAGS += -nostdlib -Wl,-gc-sections -Wl,-static
 
-ifeq (,$(filter esp_idf_heap,$(USEMODULE)))
-  # use the wrapper functions for calloc to add correct overflow detection missing
-  # in the newlib's version.
+# use the wrapper functions for calloc to add correct overflow detection missing
+# in the newlib's version.
+LINKFLAGS += -Wl,-wrap,_calloc_r
+
+ifneq (,$(filter esp_idf_heap,$(USEMODULE)))
   LINKFLAGS += -Wl,-wrap,_malloc_r
-  LINKFLAGS += -Wl,-wrap,_calloc_r
   LINKFLAGS += -Wl,-wrap,_realloc_r
   LINKFLAGS += -Wl,-wrap,_free_r
 endif
```

leandrolanzieri replied:
Wow, thanks I completely misread that conditional. I applied your suggestion in d96e4bab5f1ece133799e241f34240cff9c7467b

gschorcht replied:
The overflow detection was added by @maribu with PR #16443 :+1:. That's why `_calloc_r` wrapper function is used in any case and why I didn't realize the wrong conditional when reviewing the code :sunglasses:

leandrolanzieri commented:
@gschorcht now that ESP's implementation of log is a module on its own, does it still make sense that `esp_log_colored` is enabled when `log_color` is used? Should it just be an option of `log_esp_common`?

It seems that `esp_common_log.h` is being included directly from other places, instead of using `log.h`. Should this be changed to the common interface (that is, using `log.h` definitions)?

https://github.com/RIOT-OS/RIOT/blob/ca74a57870946670231eb12726ab5db713f45105/cpu/esp_common/Makefile.dep#L55-L57

gschorcht commented:
> Should it just be an option of `log_esp_common`?

Do you mean the change
```diff
- #if MODULE_ESP_LOG_COLORED
+ #if IS_USED(MODULE_LOG_COLOR)
```
in `cpu/esp_common/include/esp_common_log` or something else?

leandrolanzieri commented:
> > Should it just be an option of `log_esp_common`?
> 
> Do you mean the change
> 
> ```diff
> - #if MODULE_ESP_LOG_COLORED
> + #if IS_USED(MODULE_LOG_COLOR)
> ```
> 
> in `cpu/esp_common/include/esp_common_log` or something else?

No. I mean to decouple completely the dependency I linked above. Currently when `log_color` is used (which is a log backend on its own), `esp_log_colored` is pulled. IMO this should be different. When `log_esp_common` backend is chosen, this extra modules are options of it. More like it is modelled in Kconfig now:
```
menuconfig MODULE_LOG_ESP_COMMON
    bool "ESP-specific logging implementation"
    depends on HAS_ARCH_ESP

if MODULE_LOG_ESP_COMMON

config MODULE_ESP_LOG_COLORED
    bool "Use colored log output"

config MODULE_ESP_LOG_TAGGED
    bool "Add additional information to the log output"

config MODULE_ESP_LOG_STARTUP
    bool "Add additional startup information to the log output"

endif # MODULE_LOG_ESP_COMMON
```

Additionally, I noticed the esp files are using esp-specific log functions unconditionally (e.g., `LOG_TAG_ERROR`). This requires `esp_common_log.h` to do some odd things like:

```C
/**
 * Override LOG_* definitions with a tagged version. By default the function
 * name is used as tag.
 */
#ifndef MODULE_LOG_PRINTFNOFORMAT
#undef LOG_ERROR
#undef LOG_INFO
#undef LOG_WARNING
#undef LOG_DEBUG
#define LOG_ERROR(format, ...)   LOG_TAG(LOG_ERROR  , E, __func__, format, ##__VA_ARGS__)
#define LOG_WARNING(format, ...) LOG_TAG(LOG_WARNING, W, __func__, format, ##__VA_ARGS__)
#define LOG_INFO(format, ...)    LOG_TAG(LOG_INFO   , I, __func__, format, ##__VA_ARGS__)
#define LOG_DEBUG(format, ...)   LOG_TAG(LOG_DEBUG  , D, __func__, format, ##__VA_ARGS__)
#endif
```

gschorcht commented:
> No. I mean to decouple completely the dependency I linked above. Currently when `log_color` is used (which is a log backend on its own), `esp_log_colored` is pulled.

IMHO, this is correct. Either I'm completely wrong or I don't understand the problem right. One of the most important requirements of RIOT is to have the same look and feel on all platforms as well as code compatibility. Thus the user should be able to select the same log modules on all platforms. The user hasn't to know that these modules are realized by platform specific modules.

The only platform-specific options are enabled by module `esp_log_tagged` and `esp_log_startup` which enable ESP-IDF compatible log output (tagged) and additional log output from the booatloader and startup code for debugging. But all the other log modules should be enabled by the log modules that are available on all platforms and completely compatible with other platforms.

leandrolanzieri commented:
> > Additionally, I noticed the esp files are using esp-specific log functions unconditionally (e.g., `LOG_TAG_ERROR`). This requires `esp_common_log.h` to do some odd things like:
> 
> Making the log output from different sources (the bootloader, the vendor code as well as the header files and the binary libraries of the ESP-IDF) compilable on the one hand and compatible with the RIOT log module for the user on the other hand was extremely tricky.
> 
> The problem is that we have a naming conflict and log module interface incompatibilities. `LOG_*` macros are defined in RIOT's log module as well as in ESP-IDF but both define different signatures. The bootloader, the vendor code as well the header files and binary libraries in ESP-IDF expect the `LOG_*` macros as defined in ESP-IDF while the application expects the interface as defined by the RIOT log module. The goal of realizing an own log backend for ESPs was to have a compatible log interface for applications and a compatible log output no matter from where the log output comes.
> 
> To be honest, if it works as it is, we should leave it. A user who enables module `log_color` should get colored outputs, also for the log output that come from the bootloader (if `esp_log_startup` is enabled).

I did not know that all this was to fix some name collision. OK I think that for the sake of keeping the PR simple, I'll go back with the log backend changes, becayse conceptually `log_esp_common`, as is, is no backend on its own. IMO it could still use some improvements on it but we can discuss this later on a different PR.

leandrolanzieri commented:
I have dropped the changes related to the log, and also modified the related Kconfig options for esp_common. Unfortunately I have some issue with my local setup and building in docker does not work for me currently, so I could not verify that this is working with the test/log_ applications.

gschorcht commented:
> I have dropped the changes related to the log, and also modified the related Kconfig options for esp_common.

:+1:

> Unfortunately I have some issue with my local setup and building in docker does not work for me currently, so I could not verify that this is working with the test/log_ applications.

Don't worry, I will check it.

gschorcht commented:
Murdock is green :rocket: Please squash.

gschorcht commented on cpu/esp_common/Kconfig:
```diff
@@ -56,6 +56,53 @@ config HAS_ARCH_ESP
+    default y if MODULE_LOG_COLOR
+
+config MODULE_ESP_LOG_TAGGED
+    bool "Add additional information to the log output"
```
```suggestion
    bool "Add additional information to the log output"
    help
        Adds the type of the message, the system time in ms and a tag (module or function) to the output.
```

leandrolanzieri replied:
Added

gschorcht reviewed (changes requested):
One last nitpicking change request. You can squash it directly.

gschorcht commented:
Checked all `log` module combination. It works like a charm.

leandrolanzieri commented:
> Checked all `log` module combinations. It works like a charm.

Thanks a lot! I removed the CI limiting commit, let's give it a full run.

gschorcht commented:
> Thanks a lot!

I have to thank for your great work and modelling "my" CPUs and boards and for your patience :sunglasses:
